bitcoin/testBwm.c

```diff
@@ -0,0 +1,1063 @@
+//  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+//  THE SOFTWARE.
+
+#include <inttypes.h>
+#include <pthread.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <unistd.h>
+
+#include "bcash/BRBCashParams.h"
+#include "bitcoin/BRChainParams.h"
+
+#include "BRArray.h"
+#include "BRBIP39Mnemonic.h"
+#include "BRPeerManager.h"
+#include "BRTransaction.h"
+#include "BRWallet.h"
+#include "BRWalletManager.h"
+
+#ifdef __ANDROID__
+#include <android/log.h>
+#define fprintf(...) __android_log_print(ANDROID_LOG_ERROR, "bread", _va_rest(__VA_ARGS__, NULL))
+#define printf(...) __android_log_print(ANDROID_LOG_INFO, "bread", __VA_ARGS__)
+#define _va_first(first, ...) first
+#define _va_rest(first, ...) __VA_ARGS__
+#endif
+
+//
+// Helper Routine
+//
+
+static const char *
+_BRSyncModeString (BRSyncMode m) {
+    switch (m) {
+        case SYNC_MODE_BRD_ONLY:
+        return "SYNC_MODE_BRD_ONLY";
+        case SYNC_MODE_BRD_WITH_P2P_SEND:
+        return "SYNC_MODE_BRD_WITH_P2P_SEND";
+        case SYNC_MODE_P2P_WITH_BRD_SYNC:
+        return "SYNC_MODE_P2P_WITH_BRD_SYNC";
+        case SYNC_MODE_P2P_ONLY:
+        return "SYNC_MODE_P2P_ONLY";
+    }
+}
+
+static const char *
+_BRWalletManagerEventTypeString (BRWalletManagerEventType t) {
+    switch (t) {
+        case BITCOIN_WALLET_MANAGER_CONNECTED:
+        return "BITCOIN_WALLET_MANAGER_CONNECTED";
+
+        case BITCOIN_WALLET_MANAGER_CREATED:
+        return "BITCOIN_WALLET_MANAGER_CREATED";
+
+        case BITCOIN_WALLET_MANAGER_DISCONNECTED:
+        return "BITCOIN_WALLET_MANAGER_DISCONNECTED";
+
+        case BITCOIN_WALLET_MANAGER_SYNC_STARTED:
+        return "BITCOIN_WALLET_MANAGER_SYNC_STARTED";
+
+        case BITCOIN_WALLET_MANAGER_SYNC_PROGRESS:
+        return "BITCOIN_WALLET_MANAGER_SYNC_PROGRESS";
+
+        case BITCOIN_WALLET_MANAGER_SYNC_STOPPED:
+        return "BITCOIN_WALLET_MANAGER_SYNC_STOPPED";
+
+        case BITCOIN_WALLET_MANAGER_BLOCK_HEIGHT_UPDATED:
+        return "BITCOIN_WALLET_MANAGER_BLOCK_HEIGHT_UPDATED";
+    }
+    return "<BITCOIN_WALLET_MANAGER_TYPE_UNKNOWN>";
+}
+
+static const char *
+_BRWalletEventTypeString (BRWalletEventType t) {
+    switch (t) {
+        case BITCOIN_WALLET_CREATED:
+        return "BITCOIN_WALLET_CREATED";
+        case BITCOIN_WALLET_BALANCE_UPDATED:
+        return "BITCOIN_WALLET_BALANCE_UPDATED";
+        case BITCOIN_WALLET_TRANSACTION_SUBMITTED:
+        return "BITCOIN_WALLET_TRANSACTION_SUBMITTED";
+        case BITCOIN_WALLET_FEE_PER_KB_UPDATED:
+        return "BITCOIN_WALLET_FEE_PER_KB_UPDATED";
+        case BITCOIN_WALLET_FEE_ESTIMATED:
+        return "BITCOIN_WALLET_FEE_ESTIMATED";
+        case BITCOIN_WALLET_DELETED:
+        return "BITCOIN_WALLET_DELETED";
+    }
+    return "<BITCOIN_WALLET_TYPE_UNKNOWN>";
+}
+
+static int
+_isValidWalletManagerEventPair (BRWalletManagerEvent *e1, BRWalletManagerEvent *e2) {
```

Comment:
From a documentation viewpoint, this is good stuff.

Reply:
Is it worth putting this in Crypto C - essentially right next to the `enum`?

Reply:
For the String functions, I could see that. Having this valid pairs function in the library doesn't feel right.

Reply:
Understood regarding pairs of `BRCryptoWalletManagerEventType`.  Perhaps one could argue have a pairs function for `BRCryptoWalletManagerState` does make sense as: 1) it essentially define the 'state chart transitions' and 2) supports validation of `CRYPTO_WALLET_MANAGER_EVENT_CHANGED` old and new state pairs.  But, no need to slow up this PR on that.